src/PooledArrays.jl

```diff
@@ -181,93 +184,64 @@ DataAPI.refarray(pa::PooledArray) = pa.refs
-        copyto!(dest.refs, doffs, src.refs, soffs, n)
-    else
-        @inbounds for i in 0:n-1
-            dest[dstart+i] = src[sstart+i]
-        end
-    end
+if isdefined(Base, :copy!)
+    import Base.copy!
+else
+    import Future: copy!
+end
+
+# here we do not allow dest to be SubArray as copy! is intended to replace whole arrays
+# slow path will be used for SubArray
+function copy!(dest::PooledArray{T, R, N},
+               src::PooledArrOrSub{T, N, R}) where {T, N, R}
+    copy!(dest.refs, DataAPI.refarray(src))
+    src_refcount = refcount(src)
+    Threads.atomic_add!(src_refcount, 1)
+    dest.pool = DataAPI.refpool(src)
+    dest.invpool = DataAPI.invrefpool(src)
+    dest.refcount = src_refcount
     return dest
 end
 
-# TODO: handle case when dest is SubArray
-
-function Base.copyto!(dest::PooledArray{T, R, N, RA}, doffs::Union{Signed, Unsigned},
-                 src::SubArray{T, N, PooledArray{T, R, M, RA}}, soffs::Union{Signed, Unsigned},
-                 n::Union{Signed, Unsigned}) where {T, R, N, M, RA}
+function Base.copyto!(dest::PooledArrOrSub{T, N, R}, doffs::Union{Signed, Unsigned},
+                 src::PooledArrOrSub{T, N, R}, soffs::Union{Signed, Unsigned},
+                 n::Union{Signed, Unsigned}) where {T, N, R}
     n == 0 && return dest
     n > 0 || Base._throw_argerror()
-    if soffs < 1 || doffs < 1 || soffs+n-1 > length(src) || doffs+n-1 > length(dest)
+    if soffs < 1 || doffs < 1 || soffs + n - 1 > length(src) || doffs + n - 1 > length(dest)
         throw(BoundsError())
     end
 
-    src_parent = parent(src)
-
-    if length(dest.pool) == 0
-        @assert length(dest.invpool) == 0
-        Threads.atomic_add!(src_parent.refcount, 1)
-        dest.pool = src_parent.pool
-        dest.invpool = src_parent.invpool
-        Threads.atomic_sub!(dest.refcount, 1)
-        dest.refcount = src_parent.refcount
-        copyto!(dest.refs, doffs, view(src_parent.refs, src.indices), soffs, n)
-    elseif dest.pool === src.pool && dest.invpool === src.invpool
-        copyto!(dest.refs, doffs, view(src_parent.refs, src.indices), soffs, n)
+    dest_pa = dest isa PooledArray ? dest : parent(dest)
+    src_refcount = refcount(src)
+
+    # if dest_pa.pool is empty we can safely replace it as we are sure it holds
+    # no information; having this path is useful because then we can efficiently
+    # `copyto!` into a fresh `PooledArray` created using the `similar` function
+    if length(dest_pa.pool) == 0
+        @assert length(dest_pa.invpool) == 0
+        Threads.atomic_add!(src_refcount, 1)
+        dest_pa.pool = DataAPI.refpool(src)
+        dest_pa.invpool = DataAPI.invpool(src)
+        Threads.atomic_sub!(dest_pa.refcount, 1)
+        dest_pa.refcount = src_refcount
+        copyto!(DataAPI.refarray(dest), doffs, DataAPI.refarray(src), soffs, n)
+    elseif dest.pool === DataAPI.refpool(src) && dest.invpool === DataAPI.invpool(src)
```

Comment:
```suggestion
    elseif dest.pool === DataAPI.refpool(src)
        @assert dest.invpool === DataAPI.invpool(src)
```
Also shouldn't this be the first branch? If pools are `===` but empty, we shouldn't increase the refcount.

Reply:
good point - I will change it.